hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/DefaultS3ClientFactory.java

```diff
@@ -390,4 +490,63 @@ protected static AmazonS3 configureAmazonS3Client(AmazonS3 s3,
         endpoint, epr, region);
     return new AwsClientBuilder.EndpointConfiguration(endpoint, region);
   }
+
+  /**
+   * Given a endpoint string, create the endpoint URI.
+   *
+   * @param endpoint possibly null endpoint.
+   * @param conf config to build the URI from.
+   * @return an endpoint uri
+   */
+  private static URI getS3Endpoint(String endpoint, final Configuration conf) {
+
+    boolean secureConnections = conf.getBoolean(SECURE_CONNECTIONS, DEFAULT_SECURE_CONNECTIONS);
+
+    String protocol = secureConnections ? HTTPS : HTTP;
+
+    if (endpoint == null || endpoint.isEmpty()) {
+      // the default endpoint
+      endpoint = CENTRAL_ENDPOINT;
+    }
+
+    if (!endpoint.contains("://")) {
+      endpoint = String.format("%s://%s", protocol, endpoint);
+    }
+
+    try {
+      return new URI(endpoint);
+    } catch (URISyntaxException e) {
+      throw new IllegalArgumentException(e);
+    }
+  }
+
+  /**
+   * Get the bucket region.
+   *
+   * @param region AWS S3 Region set in the config. This property may not be set, in which case
+   *               ask S3 for the region.
+   * @param s3Client A S3 Client with default config, used for getting the region of a bucket.
+   * @return region of the bucket.
+   */
+  private Region getS3Region(String region, S3Client s3Client) {
```

Comment:
wondering what the behaviour here should be. If headBucket fails for reasons such as network failure, we should probably retry. If it continues to fail probably throw an error. 
returning default US_EAST_1 just means the client will be configured with us_east_1, and then when it tries to make a call eg: s3.listObjects(), this will fail as bucket is not in us_east_1 and there is no cross region access